Refactor block binding processing and attribute computation

michalczaplinski opened the pull request:
This PR refactors the processing of block bindings into 2 steps.

#### Before

`process_block_bindings()` takes the `$block_content` as input and:

1. Gets the value for each "bound" attribute from the respective source.
2. For each updated attribute calls `replace_html()` to get the updated `$block_content` filled in with the values from the updated attributes.


#### Now

`process_block_bindings()` doesn't take any input

1. Gets the value for each "bound" attribute from the respective source.
2. **Returns the updated attributes with values from the sources**.
3. The updated attributes are used in the `render()` function of the block where `replace_html()` is called with the updated attributes.



Trac ticket: https://core.trac.wordpress.org/ticket/60282

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

michalczaplinski committed:
Refactor block binding processing and attribute computation

gziolo commented on src/wp-includes/class-wp-block.php:
```diff
@@ -446,6 +447,19 @@ public function render( $options = array() ) {
+			$this->attributes = array_merge( $this->attributes, $computed_attributes );
+		}
+
+		foreach ( $computed_attributes as $attribute_name => $source_value ) {
```
This loop is only necessary when `$block_content` is not empty. Otherwise, there is nothing to replace.

michalczaplinski replied:
Makes sense. Done in 313af354adc682d85582d893439477f810e72279

gziolo commented on src/wp-includes/class-wp-block.php:
```diff
@@ -446,6 +447,19 @@ public function render( $options = array() ) {
 			}
 		}
 
+		$computed_attributes = $this->process_block_bindings();
```
It probably should be executed before all inner blocks get rendered so they have access to computed attributes through `$parent_block`.

michalczaplinski replied:
All right, I moved it to the top of render in 313af354adc682d85582d893439477f810e72279

gziolo commented on src/wp-includes/class-wp-block.php:
```diff
@@ -446,6 +447,19 @@ public function render( $options = array() ) {
+
+		// If there are computed attributes, update the attributes with the
+		// computed ones.
+		if ( ! empty( $computed_attributes ) ) {
```
What was the reason this part wasn't included in the `process_block_bindings()` method?

michalczaplinski replied:
Right! I've moved it inside of the `process_block_bindings()`

gziolo commented:
This is great as it will allow as to use the computed value from the block bindings inside block attributes passed to the inner blocks through the `$parent_block` variable and as `$this->attributes` passed to the `render_callback`. Inside the `render_callback` we also will have `$block_content` that will have the bound attributes already replaced.

michalczaplinski committed:
Refactor block attribute processing:
- wrap the replace_html with `if ( ! empty( $computed_attributes ) )`
- move the processing of block bindings to the top of render()
- move the logic to update `$this->attributes` inside of `process_block_bindings()`

michalczaplinski committed:
Update the docstring for `process_block_bindings()`

gziolo commented on src/wp-includes/class-wp-block.php:
```diff
@@ -446,6 +455,12 @@ public function render( $options = array() ) {
 			}
 		}
 
+		if ( ! empty( $computed_attributes ) ) {
```
That should ensure it only applies changes if the block's content is not empty.

```suggestion
		if ( ! empty( $computed_attributes ) && ! empty( $block_content ) ) {
```

draganescu commented:
Is there a way to opt out of replace_html and handle the new data coming from the binding in some other special way?

gziolo commented:
> Is there a way to opt out of replace_html and handle the new data coming from the binding in some other special way?

Can you elaborate a bit more on it? I don't think it's part of the changes, but it might be supported regardless with existing filters.

gziolo commented on src/wp-includes/class-wp-block.php:
```diff
@@ -228,13 +228,13 @@ public function __get( $name ) {
-	 * @param string $block_content Block content.
-	 * @param array  $block         The full block, including name and attributes.
-	 * @return string The modified block content.
+	 * @return array The modified block content.
```
```suggestion
	 * @return array The computed block attributes for the provided block bindings.
```

gziolo commented on src/wp-includes/class-wp-block.php:
```diff
@@ -411,6 +417,9 @@ public function render( $options = array() ) {
 		);
 
+		// Process the block bindings and get attributes updated with the values from the sources.
+		$computed_attributes = $this->process_block_bindings();
```
I'm comtemplating on my feedback from yesterday and started thinking whether to move back here the changes to update `$this->attributes` as it now unclear why this call is here.

michalczaplinski replied:
Yeah, I think that it's a little bit better to keep that type of side-effect
directly in the `render()` and not hide it in the method, but it's not a strong opinion.

gziolo replied:
I agree with you. Thank you for trying the alternative approach 🙇

gziolo commented:
I updated locally unit tests to validate that block attributes get updated correctly:

```diff
Index: tests/phpunit/tests/block-bindings/render.php
===================================================================
--- tests/phpunit/tests/block-bindings/render.php	(revision 57564)
+++ tests/phpunit/tests/block-bindings/render.php	(working copy)
@@ -62,6 +62,7 @@
 		$result   = $block->render();
 
 		$this->assertEquals( $expected, $result, 'The block content should be updated with the value returned by the source.' );
+		$this->assertSame( 'test source value', $block->attributes['content'] );
 	}
 
 	/**
@@ -97,5 +98,6 @@
 		$result   = $block->render();
 
 		$this->assertEquals( $expected, $result, 'The block content should be updated with the value returned by the source.' );
+		$this->assertSame( "The attribute name is 'content' and its binding has argument 'key' with value '$value'.", $block->attributes['content'] );
 	}
 }
```

We could extend the existing tests, but maybe refactor it a bit more. It would also be nice to add a test for the Image block to ensure the `url` gets correctly replaced when using the saved image placeholder which fixes the issue reported in https://github.com/WordPress/gutenberg/issues/58425:

```html
<!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image -->
```

This is the test I drafted that passes:

```php
/**
	 * Tests if the block content is updated with the value returned by the source
	 * for the Image block in the placeholder state.
	 *
	 * @ticket 60282
	 *
	 * @covers ::register_block_bindings_source
	 */
	public function test_update_block_with_value_from_source_imag_placeholder() {
		$get_value_callback = function () {
			return 'https://example.com/image.jpg';
		};

		register_block_bindings_source(
			self::SOURCE_NAME,
			array(
				'label'              => self::SOURCE_LABEL,
				'get_value_callback' => $get_value_callback,
			)
		);

		$block_content = <<<HTML
<!-- wp:image {"metadata":{"bindings":{"url":{"source":"test/source"}}}} -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image -->
HTML;

		$parsed_blocks = parse_blocks( $block_content );
		$block         = new WP_Block( $parsed_blocks[0] );

		$expected = '<figure class="wp-block-image"><img src="https://example.com/image.jpg" alt=""/></figure>';
		$result   = $block->render();

		$this->assertSame( 'https://example.com/image.jpg', $block->attributes['url'] );
		$this->assertEquals( $expected, trim( $result ), 'The block content should be updated with the value returned by the source.' );
	}
```

draganescu commented:
> Can you elaborate a bit more on it? I don't think it's part of the changes, but it might be supported regardless with existing filters.

I am aware I may be asking dumb questions 😁  - so what I meant was what if I have a block and I bind and attribute to some source but I do not want core's processing on the output. Maybe my attribute is a visibility flag and the block has other attributes too bound and I'd short-circuit the binding's replace_html and handle it in my own render callback so avoid useless server work of replacing markup for no reason.

Currently the binding process is completely on, nether `$block_content = $this->process_block_bindings( $block_content );` or `$modified_block_content = $this->replace_html` are filtered. Is it too soon to consider filters? Maybe it is.

michalczaplinski committed:
only apply changes if the block's content is not empty.

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

michalczaplinski committed:
Update docstring wording

Co-authored-by: Greg Ziółkowski <grzegorz@gziolo.pl>

michalczaplinski committed:
Merge branch 'trunk' into refactor/block-bindings-processing

michalczaplinski committed:
Add block attribute assertions in render tests

michalczaplinski committed:
Add a new unit test for rendering the Image block

michalczaplinski commented:
@gziolo I've added the unit test that you drafted, thanks! 🙂

I think that this is looking pretty good so let me know your thoughts about the https://github.com/WordPress/wordpress-develop/pull/6059#discussion_r1482695478 . I think that we can always change it in a follow-up.

gziolo reviewed (approved):
By the way, I can apply the necessary change discussed before committing tomorrow.

Great work with all refactoring this week 👏

gziolo commented:
Committed with https://core.trac.wordpress.org/changeset/57574.

mukeshpanchal27 commented on src/wp-includes/class-wp-block.php:
```diff
@@ -275,11 +274,18 @@ private function process_block_bindings( $block_content ) {
+		// If there are computed attributes, update the attributes with the
+		// computed ones.
+		if ( ! empty( $computed_attributes ) ) {
+			// Merge the computed attributes with the original attributes
```
Missing end period.

mukeshpanchal27 replied:
@gziolo ☝️

gziolo replied:
I noticed that myself, too. 🙈 I'm working on another commit and I plan to include it. Thank you as always for your attention to detail and checking new commits 🙇🏻

gziolo replied:
Done in https://github.com/WordPress/wordpress-develop/commit/acad00ecabcf44a6c4e552c4fd518ee708fd131f.

gziolo commented:
> Currently the binding process is completely on, nether $block_content = $this->process_block_bindings( $block_content ); or $modified_block_content = $this->replace_html are filtered. Is it too soon to consider filters? Maybe it is.

@draganescu, we could always discuss introducing new settings for the block binding source:
- `filter_callback` to post-process the raw value returned by the source, example: date formatting.
- `should_replace_callback` to run before `$this->replace_html()` that would let folks decide whether to replace the value for an attribute in HTML.

We should be iterating on the public API as we learn about the most common needs reported by developers.
